### PR TITLE
NOMMU/vm_mapDestroy: Remove only entries associated with the map

### PR DESCRIPTION
**Pull request by agkaminski**
DONE: RTOS-551

## Description

Fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/721

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [x] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: vm/map.c [hunk 1]
```diff
@@ -775,11 +775,29 @@ void vm_mapDestroy(process_t *p, vm_map_t *map)
...
+			/* Remove only entries associated with the map */
+			if (e->map == map) {
+				LIST_REMOVE(&p->entries, e);
+				if (next == e) {
+					next = NULL;
+				}
+				e->process = NULL;
+				lib_rbRemove(&map->tree, &e->linkage);
+				e->map = NULL;
+				map_free(e);
```
Comment (badochov) [inline vm/map.c:794]: `map_common.lock` is locked and unlocked here every time, maybe number of locking may be reduced?

File: vm/map.c [hunk 2]
```diff
@@ -775,11 +775,29 @@ void vm_mapDestroy(process_t *p, vm_map_t *map)
...
+			}
+
+			e = next;
+		} while ((e != NULL) && (e != p->entries));
 	}
+	proc_lockClear(&p->lock);
 	proc_lockClear(&map->lock);
```
Comment (badochov) [inline vm/map.c:801]: isn't `proc_lockDone(&map->lock);` missing?

File: vm/map.c [hunk 3]
```diff
@@ -775,11 +775,29 @@ void vm_mapDestroy(process_t *p, vm_map_t *map)
...
-	proc_lockSet(&map->lock);
-	while ((e = p->entries) != NULL) {
-		_map_remove(map, e);
-		map_free(e);
+	map_entry_t *next;
+
+	proc_lockSet2(&map->lock, &p->lock);
+	if (p->entries != NULL) {
+		e = p->entries;
+		do {
```
Comment (badochov) [inline vm/map.c:783]: `if` maybe removed

```c
e = p->entries;
while ((e != NULL) && (e != p->entries))
```

**Review by badochov — CHANGES_REQUESTED** [review]

File: vm/map.c (hunk 3 above)
Comment (agkaminski) [reply vm/map.c:783]: Nope, if e == p-> entries we won't enter such while

Comment (badochov) [reply vm/map.c:783]: Ow, right

File: vm/map.c (hunk 1 above)
Comment (agkaminski) [reply vm/map.c]: Done

File: vm/map.c (hunk 2 above)
Comment (agkaminski) [reply vm/map.c:819]: It's not, we never destroy a map on NOMMU for real, we only "detach" a process from it

File: vm/map.c [hunk 4]
```diff
@@ -775,11 +791,31 @@ void vm_mapDestroy(process_t *p, vm_map_t *map)
...
 #else
-	proc_lockSet(&map->lock);
-	while ((e = p->entries) != NULL) {
-		_map_remove(map, e);
-		map_free(e);
+	map_entry_t *next;
+
+	proc_lockSet2(&map->lock, &p->lock);
+	proc_lockSet(&map_common.lock);
+	if (p->entries != NULL) {
```
Comment (lukileczo) [inline vm/map.c:798]: Small optimization - locking `map_common.lock` may be moved inside the `if`

**Pull request merged**